Issue 1085 - Model/federation permissions

simplypixi opened the pull request:
This fixes #1085 

The same situation like before. I had to create this branch from #1084  branch. #1084  PR is not merged yet, but I need these changes to continue my work.

### Description
Added model's/federation's permissions list component
Added button, which redirects to "Projects permissions" view
Added footer with proper description
Added models service
Added new routes to update multiple models permissions
Added new routes to update model permissions for multiple users
Added simplebar library to achive the same scrollbar on different browsers

simplypixi commented:
The PR is ready to review. I resolved conflicts and I fixed the bug, which I found.

carmenfan commented:
@simplypixi  A few quick ones I spot since I don't think I have time to go back to checking this today:

- Can we add a message here to tell the user to select a model first? Something like this:
![image](https://user-images.githubusercontent.com/11945337/44783498-7c0b1780-ab82-11e8-8a10-4a5ab4a6f45e.png)

- Radio button needs to update here to give user feedback, but it should clear again if the model selection has changed (e.g. if the user adds another model)
![filter](https://user-images.githubusercontent.com/11945337/44783617-d2785600-ab82-11e8-8083-006448f2c48b.gif)

- Can we add that grey dash here as well when some models are checked?
![image](https://user-images.githubusercontent.com/11945337/44783654-ecb23400-ab82-11e8-89f6-ed2d680b0ad3.png)

- Project admins currently cannot edit model permissions
![filter](https://user-images.githubusercontent.com/11945337/44783762-44e93600-ab83-11e8-8b43-3ee36569d094.gif)

- The user should be able to change project in the models tab, if they do, just reset the model list with no models selected

- Currently you can get stuck if you changed teamspace in the model tab (We need to cater for project not selected on model tab):
![filter](https://user-images.githubusercontent.com/11945337/44784002-0e5feb00-ab84-11e8-8d9c-1f8c5f5aa314.gif)
):

simplypixi commented:
> Project admins currently cannot edit model permissions

@carmenfan Project admin can't do this because server returns a 401 error when app calls GET request on `/models/permissions`. It's new route, but it's similar to route for single model. For single model everything looks good. I investigated this bug, but I have no idea what's wrong. Do you have any ideas? Could you check that?

simplypixi commented:
@Charence @carmenfan I fixed the bugs and I resolved the conflicts. PR is reasy for review. ;)

simplypixi commented:
@carmenfan Thank you for detailed review!
I have a few questions to the bugs:

> Scroll is scrolling the header out on the model tab
It's visible on Windows, isn't it?

> Once you've assigned someone as project admin (and refresh the page), you can't unassign them. Any project admin should be able to unassign any project admins.
Should it work the same for teamspace admins?

Improvements/Comments:

> Can we revive the tooltip on the header ?(commenter/viewer etc)
Yes, we can.

> Can we include the model type on the search here? (i.e. if i type "model" on the search it will display everything that's not a federation)
Sure

> The bolded text on search is nice, can we have it on all the user searches as well?
Yes, definitely. I just forgot about that. :/

> Can we rearrange the tabs so it's USERS, PROJECTS then JOBS?
Yes, we can

> Is it possible to allow the user to click on the row to enable/disable check box?
I can try to do that, but it'll not be so easy, because I'll have to handle two click events in the same time. In the example aren't used interactive components - it's just a plain text.

> Can we make the dash also green?
Yes

> Can we remove check boxes on admin and replace it with an icon?
And yes ;)

carmenfan commented:
> Scroll is scrolling the header out on the model tab
It's visible on Windows, isn't it?

Yes it is still visible but it looks misaligned since there's the models list next to it! :)  On closer inspection it seems like there is 2 different set of scrolls there, can we disable the top one (or is that going to cause trouble with more data or something?)
![filter](https://user-images.githubusercontent.com/11945337/44923740-12853780-ad41-11e8-9282-1b2dc3820777.gif)


> Once you've assigned someone as project admin (and refresh the page), you can't unassign them. Any project admin should be able to unassign any project admins.
Should it work the same for teamspace admins?

Yes, so for clarity:
**Assigning/revoking Teamspace admins**
Can assign/revoke other teamspace admins, apart from;
     - You cannot revoke admin privilege from the owner (i.e. username == teamspace)
     - You cannot revoke yourself from teamspace (this is stopping user from doing something stupid and leave themselves with no admins)

**Assigning/revoking Project admins**
Any teamspace admin/project admins can assign/revoke project admin privileges.Teamspace admins are automatically project admins for all projects. 
    - You cannot revoke teamspace admin from project admin privileges.
    - You cannot revoke yourself from project admin privilege.

Hope this makes sense!

simplypixi commented:
@carmenfan I fixed the bugs and I added suggested improvements - except selection by clicking on the row. I didn't do it because is overcomplicated in AngularJS and AngularJS Material context. We can do it better using React. Is it ok for you?

simplypixi commented:
@carmenfan Could you check what I do wrong on the backend side? The tests fail. ;(

carmenfan commented:
- We need to disable model admin. It should only be there to show someone's a project admin.

- It's not selecting everyone when I click on select all (looks like it's skipping project admins?)
![filter](https://user-images.githubusercontent.com/11945337/45029050-4e6d1500-b03f-11e8-86bd-2a8fbd17b4af.gif)

- I can't unselect users
![filter](https://user-images.githubusercontent.com/11945337/45029080-680e5c80-b03f-11e8-8342-95a310f2650f.gif)

- Can we bold this one too? 
![image](https://user-images.githubusercontent.com/11945337/45029113-87a58500-b03f-11e8-8eec-1d472585b30f.png)

- Can we add the bold styling on user search?
![image](https://user-images.githubusercontent.com/11945337/45029143-9db34580-b03f-11e8-9196-03dfdc38391f.png)

- It's selecting all models when I select all
![filter](https://user-images.githubusercontent.com/11945337/45029198-ccc9b700-b03f-11e8-94f8-4ba2162b9f87.gif)

- Can we add a tooltip on the how_to_reg icon to tell the user that the entry is a teamspace admin (on project tab) / project admin (model tab)?

carmenfan commented:
> @carmenfan Could you check what I do wrong on the backend side? The tests fail. ;(

Yep, I'll do it after lunch!

simplypixi commented:
> Yep, I'll do it after lunch!
Thank you!

> Can we add a tooltip on the how_to_reg icon to tell the user that the entry is a teamspace admin (on project tab) / project admin (model tab)?
It's hard to add the tooltips here, because this component (md-tooltip) has a problem with binding to dynamically rendered components - the icon has a ngIf directive.

I fixed rest of bugs. ;)

simplypixi commented:
> Horizontal scroll should be available if the user's screen is not wide enough

But this component is in disabled state. I think, enabled scroll is not proper behavior for this state - especially that there is no data to show. Are you sure about that?

> Can we add the bold styling on user search? (from above)

This is weird. I added bolding to each search component, which I made. Maybe, we think about two different things. :D
![image](https://user-images.githubusercontent.com/5683490/45048183-872adf80-b07b-11e8-89da-271c63c02cb0.png)

> Can we make all filters case insensitive? i.e., "FAN" should be searchable using "fan", "FAN", "Fan", "FaN", etc.
Sure! ;)

Charence commented:
@simplypixi 

> Horizontal scroll should be available if the user's screen is not wide enough

> But this component is in disabled state. I think, enabled scroll is not proper behavior for this state - especially that there is no data to show. Are you sure about that?

I see! Ignore that comment then :-)

simplypixi commented:
@Charence 
> Can we add the bold styling on user search? (from above)
What about this bug? Are we thinking about the same place?

Charence commented:
@simplypixi I think we are thinking about different things. Just realised there are two issues.

1. I think "ah" should be bold and green in the same way as "ds"
![45048183-872adf80-b07b-11e8-89da-271c63c02cb0-a](https://user-images.githubusercontent.com/1438332/45089630-af125580-b104-11e8-99ed-fb7ab4b237ef.png)

2. I think the title "Model/federation" should be bold in the same style as "User"
![45048183-872adf80-b07b-11e8-89da-271c63c02cb0-b](https://user-images.githubusercontent.com/1438332/45089674-d5d08c00-b104-11e8-95f6-31adbc5b0b08.png)

simplypixi commented:
> I think "ah" should be bold and green in the same way as "ds"

Ok. It's fixed already ;)

> I think the title "Model/federation" should be bold in the same style as "User"

This title is bold (You can check in dev-tools if you want) :D Maybe it looks different for you because of the color (Active field has a "blue" label)?

simplypixi commented:
@Charence Could you check the bugs one more time? I fixed most of them on the morning. ;)
